src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs

```diff
@@ -2,74 +2,114 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 using System.Buffers.Text;
+using System.Runtime.CompilerServices;
 
 namespace System.Buffers.Reader
 {
-    public static partial class BufferReaderExtensions 
+    public ref partial struct BufferReader
     {
-        public static bool TryParse(ref BufferReader reader, out bool value) 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public unsafe bool TryParse(out bool value)
         {
-            var unread = reader.UnreadSegment;
-            if (Utf8Parser.TryParse(unread, out value, out int consumed))
+            var unread = UnreadSegment;
+            if (Utf8Parser.TryParse(unread, out value, out int consumed) && consumed < unread.Length)
             {
-                if (unread.Length > consumed)
-                {
-                    reader.Advance(consumed);
-                    return true;
-                }
+                Advance(consumed);
+                return true;
+            }
+
+            return TryParseSlow(out value);
+        }
+
+        private unsafe bool TryParseSlow(out bool value)
+        {
+            const int MaxLength = 5;
+            var unread = UnreadSegment;
+
+            if (unread.Length > MaxLength)
+            {
+                // Fast path had enough space but couldn't find valid data
+                value = default;
+                return false;
             }
 
-            Span<byte> tempSpan = stackalloc byte[5];
-            var copied = Peek(reader, tempSpan);
-            if (Utf8Parser.TryParse(tempSpan.Slice(0, copied), out value, out consumed))
+            byte* buffer = stackalloc byte[MaxLength];
+            Span<byte> tempSpan = new Span<byte>(buffer, MaxLength);
+            if (Utf8Parser.TryParse(PeekSlow(tempSpan), out value, out int consumed))
             {
-                reader.Advance(consumed);
+                Advance(consumed);
                 return true;
             }
 
             return false;
         }
 
-        public static bool TryParse(ref BufferReader reader, out int value)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public unsafe bool TryParse(out int value)
+        {
+            var unread = UnreadSegment;
+            if (Utf8Parser.TryParse(unread, out value, out int consumed) && consumed < unread.Length)
+            {
+                Advance(consumed);
+                return true;
+            }
+
+            return TryParseSlow(out value);
+        }
+
+        private unsafe bool TryParseSlow(out int value)
         {
-            var unread = reader.UnreadSegment;
-            if (Utf8Parser.TryParse(unread, out value, out int consumed))
+            const int MaxLength = 15;
```

Comment:
What happens if the buffer contains leading zeros? Would MaxLength = 15 work in that case?

Also, where is 15 coming from?

Reply:
You have to be at the start- only the user can know what is relevant. 15 is the maximum size of the incoming string in the given format.